File: Substring-with-Concatenation-of-All-Words/Substring-with-Concatenation-of-All-Words.cpp

class Solution {
public:
    vector<int> findSubstring(string S, vector<string> &L) {
        vector<int>ans;
        int n=L.size();
        if(n==0)
            return ans;
        int len=L[0].length();
        int lens=S.length();
        if(lens==0)
            return ans;
        map<string,int>words,cur;
        for(int i=0;i<n;i++)
            words[L[i]]++;
        for(int i=0;i<=(lens-n*len);i++)
        {
            cur.clear();
            int j=0;
            for( ;j<n;j++)
            {
                string temp=S.substr(i+j*len,len);
                if(words.find(temp)==words.end())
                    break;
                cur[temp]++;
                if(words[temp]<cur[temp])
                    break;
            }
            if(j==n)
                ans.push_back(i);
        }
        return ans;
    }
};


 

File: Substring-with-Concatenation-of-All-Words/Substring-with-Concatenation-of-All-Words.py


class Solution:
    # @param S, a string
    # @param L, a list of string
    # @return a list of integer
    def findSubstring(self, S, L):
        ans=[]
        n=len(L)
        if n==0:
            return ans
        wordlen=len(L[0])
        lens=len(S)
        if lens==0:
            return ans
        dict={}
        cur={}
        for i in range(n):
            if L[i] in dict:
                dict[L[i]]+=1
            else:
                dict[L[i]]=1
        for i in range(lens-n*wordlen+1):
            cur.clear()
            j=0
            while j<n:
                temp=S[i+j*wordlen:i+j*wordlen+wordlen]
                if temp not in dict:
                    break
                if temp in cur:
                    cur[temp]+=1
                else:
                    cur[temp]=1
                if dict[temp]<cur[temp]:
                    break
                j+=1
            if j==n:
                ans.append(i)
        return ans
 